### DISCUSSION TITLE
outpatient appointments

### DISCUSSION THREAD

**Discussion by rc16** [opensafely/documentation#950] [Q&A] [answered]
Hi, 

I'd like to identify consecutive outpatient appointments and the corresponding mode of appointment. I tried the method here: https://docs.opensafely.org/study-def-tricks/ where I defined a function for the date and a function for the mode of appointment but it doesn't seem to like the section defining mode of appointment as the returning variable isn't a date. 

Is it possible to identify this information or would I need to go to summary data e.g. mode of latest appointment within periods of time?

**Comment by rebkwok** [comment, marked as answer]
In order to idenitfy both the consecutive appointments and the corresponding mode, I think you'd need to define two variables for each consecutive date, one to return the date, and one to return the corresponding mode.  

Something like this should work (note there's an extra `returning` parameter in the `var_signature()` function, so that you can use the same function to return the variable for the date and the appointment type. 

```
def outpatient_appt_date_and_mode_X(name, codelist, index_date, n, return_expectations):

    def var_signature(name, codelist, on_or_after, return_expectations, returning):
        return {
            name: patients.outpatient_appointment_date(
                    with_these_treatment_function_codes=codelist,
                    ...
                    returning=returning,
                    on_or_after=on_or_after,
                    date_format="YYYY-MM-DD",
                    find_first_match_in_period=True,
                    return_expectations=return_expectations
        ),
        }
    variables = var_signature(f"{name}_date_1", codelist, index_date, return_expectations, returning="date")
    variables.update(var_signature(f"{name}_appt_type_1", codelist, index_date, return_expectations, returning="consultation_medium_used"))
    for i in range(2, n+1):
        variables.update(var_signature(f"{name}_date_{i}", codelist, f"{name}_date_{i-1} + 1 day", return_expectations, returning="date"))
        variables.update(var_signature(f"{name}_appt_type_{i}", codelist, f"{name}_date_{i-1} + 1 day", return_expectations, returning="consultation_medium_used"))
    return variables
```

**Comment by rc16** [comment]
Thank you, that  looks like what I need. Could I check what I need to put lower down in the script to return the variables. I have added the above to the top of the script:

def op_appt_dates_and_mode_X(name, index_date, n, return_expectations):

    def var_signature(name, on_or_after, return_expectations, returning):
        return {
            name: patients.outpatient_appointment_date(
                    with_these_treatment_function_codes="410",
                    returning=returning,
                    on_or_after=on_or_after,
                    date_format="YYYY-MM-DD",
                    find_first_match_in_period=True,
                    return_expectations=return_expectations
        ),
        }
    variables = var_signature(f"{name}_date_1", index_date, return_expectations, returning="date")
    variables.update(var_signature(f"{name}_appt_type_1", index_date, return_expectations, returning="consultation_medium_used"))
    for i in range(2, n+1):
        variables.update(var_signature(f"{name}_date_{i}", f"{name}_date_{i-1} + 1 day", return_expectations, returning="date"))
        variables.update(var_signature(f"{name}_appt_type_{i}", f"{name}_date_{i-1} + 1 day", return_expectations, returning="consultation_medium_used"))
    return variables

and then lower down in the study definition:

**op_appt_dates_and_mode_X(
        name="op_appt",
        on_or_after="2019-04-01",
        return_expectations="none",
        n=10
    ),

But it says  "op_appt_dates_and_mode_X() got an unexpected keyword argument 'on_or_after'" I can't figure out why?

Also, is there any documentation, or anywhere I can find out more about defining functions?

**Reply by rebkwok** [reply]
Ah, that's because the `op_appt_dates_and_mode_X()` function takes positional arguments, not keyword arguments, and you're calling it with an unknown keyword (there's a typo in the docs, which I'll go and fix now).  It is expecting 4 arguments, that can optionally be called with the names `name`, `index_date`, `n` and `return_expectations`.  So in the study definition, you can call it with either:

```
**op_appt_dates_and_mode_X(
    "op_appt",
    "2019-04-01",
    10,
    None,
),
```
or
```
**op_appt_dates_and_mode_X(
    name="op_appt",
    index_date="2019-04-01",
    n=10,
    return_expectations=None,
),
```
(Note that the order of the arguments matters in the first version, but doesn't in the second version, when you're explicitly naming the arguments.)

To use the `default_expectations` as the `return_expectations`, you need to pass the argument as `None` (not the string `"none"`) - but you could achieve the same thing by just omitting it entirely and define the function instead as:
```
def op_appt_dates_and_mode_X(name, index_date, n):
    def var_signature(name, on_or_after, returning):
        return {
            name: patients.outpatient_appointment_date(
                    with_these_treatment_function_codes="410",
                    returning=returning,
                    on_or_after=on_or_after,
                    date_format="YYYY-MM-DD",
                    find_first_match_in_period=True,
        ),
        }
    variables = var_signature(f"{name}_date_1", index_date, returning="date")
    variables.update(var_signature(f"{name}_appt_type_1", index_date, returning="consultation_medium_used"))
    for i in range(2, n+1):
        variables.update(var_signature(f"{name}_date_{i}", f"{name}_date_{i-1} + 1 day", returning="date"))
        variables.update(var_signature(f"{name}_appt_type_{i}", f"{name}_date_{i-1} + 1 day", returning="consultation_medium_used"))
    return variables
```

**Comment by rc16** [comment]
Ok, thank you very much, I've got it working.
